pkg/chains/formats/intotoite6/taskrun/material.go

```diff
@@ -0,0 +1,178 @@
+		if !isDuplicateMaterial(mats, m) {
+			mats = append(mats, m)
+		}
+	}
+
+	// add sidecar images
+	for _, sidecarState := range tro.Status.Sidecars {
+		m := imageIDToMaterial(sidecarState.ImageID)
+		if !isDuplicateMaterial(mats, m) {
+			mats = append(mats, m)
+		}
+	}
+
+	gitCommit, gitURL := gitInfo(tro)
+
+	// Store git rev as Materials and Recipe.Material
+	if gitCommit != "" && gitURL != "" {
+		mats = append(mats, slsa.ProvenanceMaterial{
+			URI:    gitURL,
+			Digest: map[string]string{"sha1": gitCommit},
+		})
+		return mats
+	}
+
+	sms := artifacts.RetrieveMaterialsFromStructuredResults(tro, artifacts.ArtifactsInputsResultName, logger)
+	mats = append(mats, sms...)
+
+	if tro.Spec.Resources == nil {
+		return mats
+	}
+	// check for a Git PipelineResource
+	for _, input := range tro.Spec.Resources.Inputs {
+		if input.ResourceSpec == nil || input.ResourceSpec.Type != v1alpha1.PipelineResourceTypeGit {
+			continue
+		}
+
+		m := slsa.ProvenanceMaterial{
+			Digest: slsa.DigestSet{},
+		}
+
+		for _, rr := range tro.Status.ResourcesResult {
+			if rr.ResourceName != input.Name {
+				continue
+			}
+			if rr.Key == "url" {
+				m.URI = attest.SPDXGit(rr.Value, "")
+			} else if rr.Key == "commit" {
+				m.Digest["sha1"] = rr.Value
+			}
+		}
+
+		var url string
+		var revision string
+		for _, param := range input.ResourceSpec.Params {
+			if param.Name == "url" {
+				url = param.Value
+			}
+			if param.Name == "revision" {
+				revision = param.Value
+			}
+		}
+		m.URI = attest.SPDXGit(url, revision)
+		mats = append(mats, m)
+	}
+
+	return mats
+}
+
+// gitInfo scans over the input parameters and looks for parameters
+// with specified names.
+func gitInfo(tro *objects.TaskRunObject) (commit string, url string) {
+	// Scan for git params to use for materials
+	if tro.Status.TaskSpec != nil {
+		for _, p := range tro.Status.TaskSpec.Params {
+			if p.Default == nil {
+				continue
+			}
+			if p.Name == attest.CommitParam {
+				commit = p.Default.StringVal
+				continue
+			}
+			if p.Name == attest.URLParam {
+				url = p.Default.StringVal
+			}
+		}
+	}
+
+	for _, p := range tro.Spec.Params {
+		if p.Name == attest.CommitParam {
+			commit = p.Value.StringVal
+			continue
+		}
+		if p.Name == attest.URLParam {
+			url = p.Value.StringVal
+		}
+	}
+
+	for _, r := range tro.Status.TaskRunResults {
+		if r.Name == attest.CommitParam {
+			commit = r.Value.StringVal
+		}
+		if r.Name == attest.URLParam {
+			url = r.Value.StringVal
+		}
+	}
+
+	url = attest.SPDXGit(url, "")
+	return
+}
+
+func imageIDToMaterial(imageID string) slsa.ProvenanceMaterial {
+	m := slsa.ProvenanceMaterial{
+		Digest: slsa.DigestSet{},
+	}
+
+	uriDigest := strings.Split(imageID, uriSeparator)
+	if len(uriDigest) == 2 {
+		digest := strings.Split(uriDigest[1], digestSeparator)
+		if len(digest) == 2 {
+			// no point in partially populating the material
+			// do it if both conditions are valid.
+			m.URI = uriDigest[0]
```

Comment:
Something to keep an eye on...

The pipeline controller doesn't include extra information about how the image was referenced, so when flattening the images down to the list of materials it becomes harder to know which image is which ([example](https://gist.github.com/wlynch/a88ffebb8487e9759eeb35360b859ced)). This is the same issue https://github.com/slsa-framework/slsa/blob/e89e1fc90273573f2d2591e621a1949e7196a930/docs/provenance/v0.1.md?plain=1#L307-L308 alludes to.

My gut feeling is you're already doing the right thing here since it normalizes the image output to make it easier to query for images, then you could use other provenance data to figure out which image ran for which step.

We should include a test for same image repo URI w/ different digest just as a precaution.

Reply:
Sounds good! I will add the unit test.